In [4]:
import numpy as np
import pandas as pd

## Definições

- Função

In [5]:
def f(x1,x2):
    return 3**(1/(x1+x2))+x1**2+x2**2

- Gradiente

In [6]:
def Gradiente(x):
    x = np.array(x)
    x1 = x[0][0]
    x2 =x[1][0]
    return np.array([[2*x1-(np.log(3)*np.exp(np.log(3)/(x1+x2)))/(x1+x2)**2],
                     [2*x2-(np.log(3)*np.exp(np.log(3)/(x1+x2)))/(x1+x2)**2]])

- Hessiana

In [7]:
def Heissiana(x):
    x = np.array(x)
    x1 = x[0][0]
    x2 =x[1][0]
    return np.array([[2 - (np.log(3)*(-np.log(3)*np.exp(np.log(3)/(x1+x2))-2*np.exp(np.log(3)/(x1+x2))))/(x1+x2)**4, -np.log(3)*(-np.log(3)*np.exp(np.log(3)/(x1+x2))-2*np.exp(np.log(3)/(x1+x2)))/(x1+x2)**4],

                     [-np.log(3)*(-np.log(3)*np.exp(np.log(3)/(x1+x2))-2*np.exp(np.log(3)/(x1+x2)))/(x1+x2)**4 ,2 - (np.log(3)*(-np.log(3)*np.exp(np.log(3)/(x1+x2))-2*np.exp(np.log(3)/(x1+x2))*(x1+x2)))/(x1+x2)**4]])

# Estudo da Função
<hr/>

## Primeira foto da função

![Geogebra function](images/function.png)



Apenas pela observação direta da função podemos visualizar criada pelo geogebra:
1. Há uma uma descontinuidade na função;
2. Há um mínimo local no primeiro quadrante (x > 0 e y > 0);
3. No domínio observado, a função é convexa.


## Decontinuidade

Devido a razão $\frac{1}{x_1 + x_2}$ presente no exponente do 3, temos que:

$$
\lim_{x_1 \to 0} f(x_1, x_2) = \infty \text{  e  } \lim_{x_2 \to 0} f(x_1, x_2) = \infty
$$

Isso significa que a função não está definida no ponto (0, 0) e, portanto, ele não pode ser o mínimo global da função como aparenta na imagem.\\


## Pontos estacionários

Para calcular os pontos estácionários da função, vamos igualar seu gradiante a zero

$$
\nabla f(x_1, x_2) = \begin{bmatrix}
        2x - \frac{(\ln{3}e^{\frac{\ln{3}}{x + y}})}{(x + y)^2}\\
        2y - \frac{(\ln{3}e^{\frac{\ln{3}}{x + y}})}{(x + y)^2}
\end{bmatrix} = 0
$$

Assim, obtemos um sistema de equações que pode ser resolvido subtraindo a primeira equação da segunda:

$$
        2x - 2y =  0\\
        x = y
$$

Isso indica que os pontos estacionários estão no domnínio estabelecido pela reta x = y. <br>
Vamos calcular e visualizar o que acontece quando o hiperplano formado pela reta intercepta a função.

$$
        g(x) = f(x, x) = 3^{\frac{1}{2x}} + 2x^2
$$


Intersseção em 3D            |  Intersseção em 2D (curva formada no eixo z e no eixo x)
:-------------------------:|:-------------------------:
![](images/hiperplano.png)  |  ![](images/intersse.png)



Conseguimos observar o <strong>mínimo local</strong> mencionado anteriormente e aproximação feita pelo Geogebra do seu valor, mas gostaríamos de calcular esse valor numericamente. <br>Como não foi possível terminar a resolução do sistema de equações acima com os métodos tradicionais, optamos por usar uma <strong>aproximação de Taylor de terceira ordem </strong> para aproximar a função $3^{\frac{1}{2x}}$ na forma de um polinômio.


 - Série de Taylor desenvoldida:
$$
        3^{\frac{1}{2x}} \approx \sum_{n = -3}^0 \frac{x^n (\frac{2}{\ln{3}})^n}{(-n!)} = 1\:+\:\frac{0.5493061443340549}{x}\:+\:\frac{0.15086862010157276}{x^2}+\frac{0.02762435333633141}{x^3}
$$
<p align="center">
  <img src="images/taylo.png" />
</p>




In [13]:
def CoeffTaylor(n):
    return (2/np.log(3))**n/(np.math.factorial(-n))
CoeffTaylor(0), CoeffTaylor(-1), CoeffTaylor(-2), CoeffTaylor(-3)

(1.0, 0.5493061443340549, 0.15086862010157276, 0.02762435333633141)


Dessa forma, foi possível encontrar um valor aproximadO para o ponto crítico da função algebricamente:

$$
        
        g'(x) = 4x-\frac{0.5493}{x^2}-\frac{0.30172}{x^3}-\frac{0.08286}{x^4}\\
         4x-\frac{0.5493}{x^2}-\frac{0.30172}{x^3}-\frac{0.08286}{x^4} = 0\\
        \therefore x = 0.66714
$$
O ponto estacionário encontrado foi ~(0.66714, 0.66714) com f(x, y) = 16832, que é muito parecido com o ponto aproximado pelo software ~(0.67633,0.67633) com f(x, y) = 3.16767


## Convexidade

Para analisar a convexidade vamos avaliar a Hessiana da função:

Seja w = $\frac{\ln{(3)} (\ln{(3)} e^{\frac{\ln{3}}{x+y}} + 2 (x + y) e^{\frac{\ln{3}}{x+y}})}{(x+y)^4}$, a Hessiana pode ser escrita da seguinte forma:


$$
H = \begin{bmatrix}
2+w & w \\
w & 2+w 
\end{bmatrix}
$$
Sabemos que para um matriz ser definida positiva, a sua determinante e a determinante de suas submatrices precisa ser maior que zero:


$$
    det([2+w]) = 2 + w > 0 \\
    det(H) = (2+w)^2 - w^2 > 0
$$

Se x > 0 e y > 0, w é sempre postivo o que torna as duas desigualdades acima verdadeiras. Então, a matriz é definida positiva nesse domínio e portanto convexa. Isso já indica que nosso ponto estacionário é um ponto de mínimo, vamos conferir:




In [15]:
def definida_positiva_2x2(matrix):    
    if matrix[0][0] > 0 and np.linalg.det(matrix)>0:
        return True
    else: 
        return False
    
heissiana_ponto_critico = Heissiana([[0.66714],[0.66714]])
definida_positiva_2x2(heissiana_ponto_critico)

True

Provamos, então, que o ponto crítico encontrado nesta análise é um ponto de <strong>mínimo</strong>. <br>
É fácil perceber que ele é <strong>local</strong>, pois existem valores em que a função é menor.


In [17]:
f([[-1], [-1]])

2.5773502691896257